### ISSUE TITLE
warning 6133 when namespace declaration follows import

### ISSUE THREAD

**Issue by JirkaDellOro** [microsoft/TypeScript#33317]
**tsconfig.json**: "noUnusedLocals": true,       
**tsc-version**: Version 3.6.2

```typescript
import * as Http from "http";
import * as Url from "url";

namespace Test {
 ...
}
```

**Result**: Warning 6133 'Test' is declared but its value is never read.
**Expected**: No Warning

**Comment by AnyhowStep** [comment]
You're not exporting the namespace.
Are you using it inside the file?

**Comment by JirkaDellOro** [comment]
I'm using the namespace solely for separation of, well, namespaces. It's not about building a module. So I don't think exporting it would make sense.

**Comment by AnyhowStep** [comment]
Are you referencing `Test.someExportedName` inside the file?

**Comment by JirkaDellOro** [comment]
No, it shouldn't be necessary to address anything inside the namespace explicitely with the full identifier including the namespace. I don't have exported members inside the namespace also, and it wouldn't change the behaviour. At this point of time I can either add

- //@ts-ignore no-unused-variable
- export to the namespace

to suppress the behaviour, both of which are not very elegant.

**Comment by AnyhowStep** [comment]
The `namespace` is not necessary, then.

Your file itself acts as its own namespace...

**Comment by fatcerberus** [comment]
Like @AnyhowStep said, if you're using ES modules (which you are since your example has `import`) then you don't need the namespace - your declarations won't leak outside the file you declared them in.  It's different from the olden days where everything was global by default.

**Comment by fatcerberus** [comment]
@JerkaDellOro It’s fairly likely the error is alerting you to a larger issue: because you’ve added `import` statements, TS now sees the file as an ES module and therefore *nothing in `Test` is accessible from any other file* unless you export it.  It’s therefore dead code and the compiler is warning you about it.

**Comment by fatcerberus** [comment]
`import` being present means the file is an ES module, by definition.  You can’t use that syntax in normal scripts, and there’s no way to transpile it to one in a way that wouldn’t violate the ES specification.

**Comment by AnyhowStep** [comment]
https://github.com/microsoft/TypeScript/issues/30994#issuecomment-492017219

**Comment by fatcerberus** [comment]
To be clear: ESM, at runtime, is a *separate execution model* compared to execution of regular scripts, and `import` is only supported when the engine is running in the former mode.  You thus can't get away from the scoping rules when using `import`, TS would just introduce runtime errors if it were to ignore them.

**Comment by JirkaDellOro** [comment]
@AnyhowStep: Thanks for pointing me there. In the discussion, the person you cite also stated this:
>one of the major features of namespaces in the traditional sense is cross-file scope merging,   

That is one of the reasons I use it, it's not about building a module. I'd use `module` then.

<hr>

@fatcerberus: 
> import being present means the file is an ES module, by definition. You can’t use that syntax in normal scripts, and there’s no way to transpile it to one in a way that wouldn’t violate the ES specification.  

Would you mind to point me to that specification? I can't find it stated like this in the official specs. Also the compiled code (using ts-ignore) actually works fine and as expected, it's just that the compiler warning doesn't seem to fit or is needless.

**Comment by JirkaDellOro** [comment]
@RyanCavanaugh: Thanks for tuning in.

So I understand the intended behaviour must be to print a warning that the namespace is not read, when its contents accesses imported modules. I can't grasp the idea behind that intension, though. Why does using imports change the meaning or behaviour of the namespace? There is no warning without imports and the compiled code works as intended also with the imports.

I'd love to have an explanation I can tell my students so that they know why they have to use `export` or disable the compilerwarning.

**Comment by JirkaDellOro** [comment]
@RyanCavanaugh ?

**Comment by JirkaDellOro** [comment]
> To be clear: ESM, at runtime, is a _separate execution model_ compared to execution of regular scripts, and `import` is only supported when the engine (or transpiler) is running in the former mode. You thus can't get away from the local-scoping rules when using `import`, TS would just introduce runtime errors if it were to ignore them.

@fatcerberus : Do you know where I can look that up?

**Comment by fatcerberus** [comment]
The pain due to ESM being a separate execution model from normal scripts is why we ended up with `.mjs` in Node and `type=module` in browsers.  It's all very well documented at this point with just a quick Google search.

But here's what the ECMAScript spec has to say anyway:
* https://www.ecma-international.org/ecma-262/#sec-ecmascript-language-scripts-and-modules

Notice that only the Module grammar allows `ImportDeclaration` and `ExportDeclaration` productions.  Therefore when your script contains `import` TS automatically switches to Module mode because that's the only valid interpretation: it'd be a syntax error at runtime if you tried to run that as a script.  And modules have their own top-level scope.

If that `Test` namespace is somehow accessible from other files despite the `import` statements, then either:
1. Your runtime environment is not spec-compliant, OR
2. The `import`s happen to have been elided by TS because they aren't used in the code

The `"module"` option in tsconfig, to be clear, only affects what TS generates as *output*.  It doesn't affect the semantics of the code you write.  TS follows the ES specification as closely as possible in the untranspiled code, and ES says `import` = module.

**Comment by JirkaDellOro** [comment]
@fatcerberus Thank you very much for your efforts, that gives me some hints to come up with an explanation for my students.  I had to decide which workaround to recommend  
> At this point of time I can either add
>- //@ts-ignore no-unused-variable
>- export to the namespace        
>
>to suppress the behaviour, both of which are not very elegant.

I go with export now and stick to "by definition", though it still should work and works to use a module in global code without exporting. At least, export does no harm...
